// resium/dist/resium.js
import { createContext as e, forwardRef as t, useCallback as n, useContext as r, useEffect as i, useImperativeHandle as a, useLayoutEffect as o, useMemo as s, useRef as c, useState as l } from "react";
import { BillboardCollection as u, BillboardGraphics as d, BoxGraphics as f, BufferPoint as p, BufferPointCollection as ee, BufferPolygon as m, BufferPolygonCollection as h, BufferPolyline as g, BufferPolylineCollection as te, Cesium3DTilesTerrainProvider as _, Cesium3DTileset as v, Cesium3DTilesetGraphics as y, CesiumWidget as b, ClassificationPrimitive as x, CloudCollection as S, CorridorGraphics as C, CubeMapPanorama as w, CustomDataSource as T, CylinderGraphics as E, CzmlDataSource as D, EllipseGraphics as O, EllipsoidGraphics as k, Entity as A, EquirectangularPanorama as j, Event as ne, Fog as M, GeoJsonDataSource as N, Google2DImageryProvider as P, GroundPolylinePrimitive as re, GroundPrimitive as ie, ImageryLayer as ae, KmlDataSource as oe, LabelCollection as se, LabelGraphics as ce, Model as le, ModelGraphics as ue, Moon as de, ParticleSystem as fe, PathGraphics as pe, PlaneGraphics as me, PointGraphics as he, PointPrimitiveCollection as ge, PolygonGraphics as _e, PolylineCollection as ve, PolylineGraphics as ye, PolylineVolumeGraphics as be, PostProcessStage as xe, PostProcessStageComposite as Se, PostProcessStageLibrary as F, Primitive as Ce, RectangleGraphics as we, SceneMode as I, ScreenSpaceEventHandler as L, ScreenSpaceEventType as R, Sun as z, TimeDynamicPointCloud as Te, Viewer as Ee, WallGraphics as De, WebMapTileServiceImageryProvider as Oe } from "cesium";
import { Fragment as B, jsx as V } from "react/jsx-runtime";
import { createPortal as ke } from "react-dom";
//#region src/core/context.ts
var H = e({}), { Provider: Ae, Consumer: U } = H, W = () => r(H) || {}, G = (e, t) => {
	let n = (e) => {
		let n = W(), r = c(void 0), a = c(!1);
		return i(() => () => {
			n.camera && e.cancelFlightOnUnmount && n.camera.cancelFlight();
		}, [n.camera, e.cancelFlightOnUnmount]), i(() => {
			n.camera && n.scene && !n.scene.isDestroyed() && (!e.once || !a.current) && (n.camera.cancelFlight(), t(n.camera, e, r.current), a.current = !0), r.current = e;
		}), null;
	};
	return n.displayName = e, n;
};
//#endregion
//#region src/core/util.ts
function je(e, t) {
	return t ? K(e).reduce((e, [n, r]) => (q(t, n) && (e[n] = r), e), {}) : {};
}
function K(e) {
	return Object.keys(e).map((t) => [t, e[t]]);
}
function q(e, t) {
	return !!e && e.indexOf(t) !== -1;
}
function Me(e, t) {
	return !!e && !!t && [...Object.keys(e), ...Object.keys(t)].every((n) => e[n] === t[n]);
}
function Ne(e) {
	return e && typeof e.isDestroyed == "function" && typeof e.destroy == "function";
}
function Pe(e) {
	return Ne(e) && e.isDestroyed();
}
function J(e) {
	return e && typeof e == "object" && "then" in e && typeof e.then == "function";
}
//#endregion
//#region src/core/EventManager.ts
var Y = "__RESIUM_EVENT_MANAGER", X = [
	"onClick",
	"onDoubleClick",
	"onMouseDown",
	"onMouseUp",
	"onMiddleClick",
	"onMiddleDown",
	"onMiddleUp",
	"onMouseMove",
	"onPinchEnd",
	"onPinchMove",
	"onPinchStart",
	"onRightClick",
	"onRightDown",
	"onRightUp",
	"onWheel",
	"onMouseEnter",
	"onMouseLeave"
], Z = class e {
	static eventTypeMap = {
		onClick: R.LEFT_CLICK,
		onDoubleClick: R.LEFT_DOUBLE_CLICK,
		onMouseDown: R.LEFT_DOWN,
		onMouseUp: R.LEFT_UP,
		onMiddleClick: R.MIDDLE_CLICK,
		onMiddleDown: R.MIDDLE_DOWN,
		onMiddleUp: R.MIDDLE_UP,
		onMouseMove: R.MOUSE_MOVE,
		onPinchEnd: R.PINCH_END,
		onPinchMove: R.PINCH_MOVE,
		onPinchStart: R.PINCH_START,
		onRightClick: R.RIGHT_CLICK,
		onRightDown: R.RIGHT_DOWN,
		onRightUp: R.RIGHT_UP,
		onWheel: R.WHEEL,
		onMouseEnter: R.MOUSE_MOVE,
		onMouseLeave: R.MOUSE_MOVE
	};
	scene;
	sshe;
	events = {
		onClick: /* @__PURE__ */ new Map(),
		onDoubleClick: /* @__PURE__ */ new Map(),
		onMouseDown: /* @__PURE__ */ new Map(),
		onMouseUp: /* @__PURE__ */ new Map(),
		onMiddleClick: /* @__PURE__ */ new Map(),
		onMiddleDown: /* @__PURE__ */ new Map(),
		onMiddleUp: /* @__PURE__ */ new Map(),
		onMouseMove: /* @__PURE__ */ new Map(),
		onPinchEnd: /* @__PURE__ */ new Map(),
		onPinchMove: /* @__PURE__ */ new Map(),
		onPinchStart: /* @__PURE__ */ new Map(),
		onRightClick: /* @__PURE__ */ new Map(),
		onRightDown: /* @__PURE__ */ new Map(),
		onRightUp: /* @__PURE__ */ new Map(),
		onWheel: /* @__PURE__ */ new Map(),
		onMouseEnter: /* @__PURE__ */ new Map(),
		onMouseLeave: /* @__PURE__ */ new Map()
	};
	hovered = void 0;
	constructor(e) {
		this.scene = e, this.sshe = new L(e?.canvas);
	}
	destroy() {
		this.hovered = void 0, this.sshe.isDestroyed() || this.sshe.destroy();
	}
	isDestroyed() {
		return this.sshe.isDestroyed();
	}
	on(e, t, n) {
		e && t === "onWheel" || this.events[t].set(e, n);
	}
	off(e, t) {
		this.events[t].delete(e), this.hovered === e && (this.hovered = void 0);
	}
	setEvents(e, t) {
		K(t).forEach(([t, n]) => {
			let r = t;
			q(X, r) && (n ? this.on(e, r, n) : this.off(e, r));
		}), this.commit();
	}
	clearEvents(e) {
		this.hovered = void 0, X.forEach((t) => {
			this.off(e, t);
		}), this.commit();
	}
	commit() {
		let t = this.sshe, n = this.sshe.isDestroyed();
		n || (this.events.onMouseEnter.size === 0 && this.events.onMouseLeave.size === 0 && this.events.onMouseMove.size === 0 ? this.sshe.removeInputAction(R.MOUSE_MOVE) : this.sshe.getInputAction(R.MOUSE_MOVE) || this.sshe.setInputAction(this.onMouseMove, R.MOUSE_MOVE)), K(this.events).forEach(([r, i]) => {
			if (r === "onMouseEnter" || r === "onMouseLeave" || r === "onMouseMove") return;
			let a = e.eventTypeMap[r];
			n || (i.size === 0 ? t.removeInputAction(a) : t.getInputAction(a) || t.setInputAction(this.eventCallback(r), a));
		});
	}
	getScreenSpaceEventHandler() {
		return this.sshe;
	}
	getEventCallback(e, t) {
		return t === null ? this.events[e].get(null) : this.events[e].get(t.id) || this.events[e].get(t.id?.entityCollection?.owner) || this.events[e].get(t.primitive) || this.events[e].get(t.tileset);
	}
	onMouseMove = (e) => {
		let t = this.pick(e.endPosition);
		this.hovered !== t && (this.hovered && (this.getEventCallback("onMouseLeave", this.hovered)?.(e, this.hovered), this.getEventCallback("onMouseLeave", null)?.(e, this.hovered)), t && (this.getEventCallback("onMouseEnter", t)?.(e, t), this.getEventCallback("onMouseEnter", null)?.(e, t))), t && this.getEventCallback("onMouseMove", t)?.(e, t), this.getEventCallback("onMouseMove", null)?.(e, t), this.hovered = t;
	};
	eventCallback = (e) => (t) => {
		let n = this.pick(t?.position);
		n && this.getEventCallback(e, n)?.(t, n), this.getEventCallback(e, null)?.(t, n);
	};
	pick(e) {
		if (e) return this.scene?.pick(e);
	}
}, Fe = ({ name: e, create: t, destroy: r, provide: s, update: u, cesiumReadonlyProps: d, cesiumEventProps: f, otherProps: p, setCesiumPropsAfterCreate: ee, useCommonEvent: m, useRootEvent: h }, g, te) => {
	let _ = c(void 0), v = W(), y = c(s ? {} : void 0), b = c({}), x = c(Ie(g)), S = c({}), [C, w] = l(!1), T = c(!1), E = c(null), D = c(void 0), O = v?.[Y], k = c(void 0), A = c(void 0), j = n(async (t) => {
		if (!_.current) return;
		let n = _.current, r = Object.keys(t), i = Object.keys(f || []), a = r.concat(Object.keys(S.current).filter((e) => !r.includes(e))).filter((e) => S.current[e] !== t[e]).map((e) => [
			e,
			S.current[e],
			t[e]
		]), o = [];
		for (let [e, t, r] of a) if (d?.includes(e)) o.push(e);
		else if (q(i, e)) {
			let i = f?.[e], a = n[i];
			a instanceof ne && (t === void 0 ? (a.addEventListener(r), b.current[i] = r) : r === void 0 ? (a.removeEventListener(t), delete b.current[i]) : (a.removeEventListener(t), a.addEventListener(r)));
		} else e !== "children" && !X.includes(e) && !p?.includes(e) && (n[e] = r);
		let s = h ? y.current?.[Y] : O;
		if (m && s && _.current && s.setEvents(h ? null : _.current, t), u && T.current) {
			let e = u(_.current, t, S.current, v);
			J(e) && await e;
		}
		S.current = t, x.current = t, T.current && o.length > 0 && (process.env.NODE_ENV !== "production" && console.warn(`Warning: <${e}> is recreated because following read-only props have been updated: ${o.join(", ")}`), N(), await P(), k.current = M());
	}, []), M = n(async () => {
		await new Promise((e) => queueMicrotask(() => e(void 0)));
		let e = t?.(v, x.current, E.current), n;
		if (n = J(e) ? await e : e, Array.isArray(n) ? (_.current = n[0], D.current = n[1]) : _.current = n, ee) await j(x.current);
		else {
			if (_.current && f) {
				let e = _.current;
				for (let t of Object.keys(x.current)) {
					let n = f[t];
					if (n) {
						let r = x.current[t], i = e[n];
						r && i instanceof ne && i.addEventListener(r);
					}
				}
			}
			S.current = x.current;
		}
		s && _.current && (y.current = {
			...v,
			...s(_.current, v, g, D.current)
		});
		let r = h ? y.current?.[Y] : O;
		m && r && _.current && r.setEvents(h ? null : _.current, x.current), A.current || w(!0);
	}, []), N = n(() => {
		w(!1), T.current = !1;
	}, []), P = n(async () => {
		await new Promise((e) => queueMicrotask(() => e(void 0))), k.current &&= (await k.current, void 0), _.current && r && r(_.current, v, E.current, D.current);
		let e = h ? y.current?.[Y] : O;
		if (m && e && _.current && e.clearEvents(h ? null : _.current), _.current && !Pe(_.current)) {
			let e = Object.keys(b.current);
			for (let t of e) _.current[t]?.removeEventListener?.(b.current[t]);
		}
		b.current = {}, y.current = void 0, D.current = void 0, _.current = void 0;
	}, []);
	return o(() => ((async () => {
		A.current &&= (await A.current, void 0), k.current = M();
	})(), () => {
		N(), A.current = P();
	}), [
		M,
		P,
		N
	]), i(() => {
		(async () => {
			k.current && await k.current;
			let e = Ie(g);
			C ? Me(e, S.current) || (await j(e), v.__$internal?.onUpdate?.()) : (S.current = e, x.current = e, T.current = !0);
		})();
	}, [
		v.__$internal,
		C,
		g,
		j
	]), a(te, () => ({ cesiumElement: C ? _.current : null }), [C]), [
		y.current,
		C,
		E
	];
};
function Ie(e) {
	let { children: t, ...n } = e;
	return n;
}
//#endregion
//#region src/core/component.tsx
var Q = ({ renderContainer: e, noChildren: n, containerProps: r, defaultProps: i, ...a }) => {
	let o = (t, o) => {
		let s = {
			...i,
			...t
		}, [c, l, u] = Fe(a, s, o);
		if (n) return null;
		let d = l && "children" in s ? s.children : null, f = e ? /* @__PURE__ */ V("div", {
			"data-testid": "resium-container",
			ref: u,
			...typeof r == "function" ? r(s) : je(s, r),
			children: d
		}) : d ? /* @__PURE__ */ V(B, { children: d }) : null;
		return c ? /* @__PURE__ */ V(H.Provider, {
			value: c,
			children: f
		}) : f;
	};
	return o.displayName = a.name, t(o);
}, Le = ["enabled", "selected"], $ = (e) => Q({
	name: e.name,
	create(t, n) {
		if (!t.scene) return;
		let r = e.create(n, t.scene.postProcessStages);
		return typeof n.enabled == "boolean" && (r.enabled = n.enabled), n.selected && "selected" in r && (r.selected = n.selected), e.props.forEach((t) => {
			!q(e.readonlyProps, t) && n[t] !== void 0 && (r.uniforms[t] = n[t]);
		}), !e.noMount && t.scene && !t.scene.isDestroyed() && t.scene.postProcessStages.add(r), r;
	},
	destroy(t, n) {
		e.noMount ? t.enabled = !1 : (n.scene && !n.scene.isDestroyed() && n.scene.postProcessStages.remove(t), t.isDestroyed() || t.destroy());
	},
	update(t, n, r) {
		e.props.forEach((i) => {
			!q(e.readonlyProps, i) && n[i] !== r[i] && (t.uniforms[i] = n[i]);
		});
	},
	cesiumProps: Le,
	cesiumReadonlyProps: e.readonlyProps,
	defaultProps: { enabled: !0 }
}), Re = (e) => t((t, n) => /* @__PURE__ */ V(U, { children: (r) => /* @__PURE__ */ V(e, {
	...t,
	ref: n,
	cesium: r
}) })), ze = Q({
	name: "Billboard",
	create(e, t) {
		return e.billboardCollection?.add(t);
	},
	destroy(e, t) {
		t.billboardCollection && !t.billboardCollection.isDestroyed() && t.billboardCollection.remove(e);
	},
	cesiumProps: [
		"alignedAxis",
		"color",
		"disableDepthTestDistance",
		"distanceDisplayCondition",
		"eyeOffset",
		"height",
		"heightReference",
		"horizontalOrigin",
		"image",
		"pixelOffset",
		"pixelOffsetScaleByDistance",
		"position",
		"rotation",
		"scale",
		"scaleByDistance",
		"show",
		"sizeInMeters",
		"splitDirection",
		"translucencyByDistance",
		"verticalOrigin",
		"width",
		"id"
	],
	useCommonEvent: !0
}), Be = Q({
	name: "BillboardCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new u({
			modelMatrix: t.modelMatrix,
			debugShowBoundingVolume: t.debugShowBoundingVolume,
			scene: e.scene,
			blendOption: t.blendOption
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { billboardCollection: e };
	},
	cesiumProps: [
		"blendOption",
		"coarseDepthTestDistance",
		"debugShowBoundingVolume",
		"debugShowTextureAtlas",
		"modelMatrix",
		"show",
		"threePointDepthTestDistance"
	],
	setCesiumPropsAfterCreate: !0
}), Ve = Q({
	name: "BillboardGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new d(t);
		return e.entity.billboard = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.billboard = void 0);
	},
	cesiumProps: [
		"image",
		"show",
		"scale",
		"horizontalOrigin",
		"verticalOrigin",
		"eyeOffset",
		"pixelOffset",
		"rotation",
		"alignedAxis",
		"width",
		"height",
		"color",
		"scaleByDistance",
		"translucencyByDistance",
		"pixelOffsetScaleByDistance",
		"imageSubRegion",
		"sizeInMeters",
		"heightReference",
		"distanceDisplayCondition",
		"disableDepthTestDistance",
		"splitDirection"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), He = Q({
	name: "BoxGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new f(t);
		return e.entity.box = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.box = void 0);
	},
	cesiumProps: [
		"heightReference",
		"dimensions",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Ue = Q({
	name: "BufferPoint",
	create(e, t) {
		if (!e.bufferPointCollection) return;
		let n = new p(), r = e.bufferPointCollection.add({
			show: t.show,
			position: t.position,
			material: t.material
		}, n);
		return t.featureId !== void 0 && (r.featureId = t.featureId), r;
	},
	destroy(e) {
		e.show = !1;
	},
	update(e, t, n) {
		t.position !== n.position && t.position !== void 0 && e.setPosition(t.position), t.material !== n.material && t.material !== void 0 && e.setMaterial(t.material);
	},
	cesiumProps: ["show", "featureId"],
	otherProps: ["position", "material"]
}), We = Q({
	name: "BufferPointCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ee({
			primitiveCountMax: t.primitiveCountMax,
			modelMatrix: t.modelMatrix
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { bufferPointCollection: e };
	},
	cesiumProps: ["show", "debugShowBoundingVolume"],
	cesiumReadonlyProps: ["primitiveCountMax", "modelMatrix"],
	setCesiumPropsAfterCreate: !0
}), Ge = Q({
	name: "BufferPolygon",
	create(e, t) {
		if (!e.bufferPolygonCollection) return;
		let n = new m(), r = e.bufferPolygonCollection.add({
			show: t.show,
			positions: t.positions,
			holes: t.holes,
			triangles: t.triangles,
			material: t.material
		}, n);
		return t.featureId !== void 0 && (r.featureId = t.featureId), r;
	},
	destroy(e) {
		e.show = !1;
	},
	update(e, t, n) {
		t.positions !== n.positions && t.positions !== void 0 && e.setPositions(t.positions), t.holes !== n.holes && t.holes !== void 0 && e.setHoles(t.holes), t.triangles !== n.triangles && t.triangles !== void 0 && e.setTriangles(t.triangles), t.material !== n.material && t.material !== void 0 && e.setMaterial(t.material);
	},
	cesiumProps: ["show", "featureId"],
	otherProps: [
		"positions",
		"holes",
		"triangles",
		"material"
	]
}), Ke = Q({
	name: "BufferPolygonCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new h({
			primitiveCountMax: t.primitiveCountMax,
			vertexCountMax: t.vertexCountMax,
			holeCountMax: t.holeCountMax,
			triangleCountMax: t.triangleCountMax,
			positionDatatype: t.positionDatatype,
			allowPicking: t.allowPicking,
			modelMatrix: t.modelMatrix
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { bufferPolygonCollection: e };
	},
	cesiumProps: ["show", "debugShowBoundingVolume"],
	cesiumReadonlyProps: [
		"primitiveCountMax",
		"vertexCountMax",
		"holeCountMax",
		"triangleCountMax",
		"positionDatatype",
		"allowPicking",
		"modelMatrix"
	],
	setCesiumPropsAfterCreate: !0
}), qe = Q({
	name: "BufferPolyline",
	create(e, t) {
		if (!e.bufferPolylineCollection) return;
		let n = new g(), r = e.bufferPolylineCollection.add({
			show: t.show,
			positions: t.positions,
			material: t.material
		}, n);
		return t.featureId !== void 0 && (r.featureId = t.featureId), r;
	},
	destroy(e) {
		e.show = !1;
	},
	update(e, t, n) {
		t.positions !== n.positions && t.positions !== void 0 && e.setPositions(t.positions), t.material !== n.material && t.material !== void 0 && e.setMaterial(t.material);
	},
	cesiumProps: ["show", "featureId"],
	otherProps: ["positions", "material"]
}), Je = Q({
	name: "BufferPolylineCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new te({
			primitiveCountMax: t.primitiveCountMax,
			vertexCountMax: t.vertexCountMax,
			modelMatrix: t.modelMatrix
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { bufferPolylineCollection: e };
	},
	cesiumProps: ["show", "debugShowBoundingVolume"],
	cesiumReadonlyProps: [
		"primitiveCountMax",
		"vertexCountMax",
		"modelMatrix"
	],
	setCesiumPropsAfterCreate: !0
}), Ye = Q({
	name: "Camera",
	create: (e) => e.scene?.camera,
	cesiumProps: [
		"position",
		"direction",
		"up",
		"right",
		"frustum",
		"defaultMoveAmount",
		"defaultLookAmount",
		"defaultRotateAmount",
		"defaultZoomAmount",
		"constrainedAxis",
		"maximumZoomFactor",
		"percentageChanged"
	],
	cesiumEventProps: {
		onChange: "changed",
		onMoveEnd: "moveEnd",
		onMoveStart: "moveStart"
	},
	setCesiumPropsAfterCreate: !0
}), Xe = G("CameraFlyHome", (e, { duration: t }) => {
	e.flyHome(t);
}), Ze = G("CameraFlyTo", (e, { onComplete: t, onCancel: n, ...r }) => {
	e.flyTo({
		...r,
		complete: t,
		cancel: n
	});
}), Qe = G("CameraLookAt", (e, { target: t, offset: n }) => {
	e.lookAt(t, n);
}), $e = G("CameraFlyToBoundingSphere", (e, { boundingSphere: t, onComplete: n, onCancel: r, ...i }) => {
	e.flyToBoundingSphere(t, {
		...i,
		complete: n,
		cancel: r
	});
}), et = Q({
	name: "Cesium3DTileset",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let n = t.url, r;
		r = J(n) ? await n : n;
		let i;
		try {
			i = await v.fromUrl(r, t), t.onReady?.(i);
		} catch (e) {
			t.onError?.(e);
			return;
		}
		return t.colorBlendAmount && (i.colorBlendAmount = t.colorBlendAmount), t.colorBlendMode && (i.colorBlendMode = t.colorBlendMode), t.style && (i.style = t.style), e.primitiveCollection.add(i), i;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: /* @__PURE__ */ "show.modelMatrix.shadows.maximumScreenSpaceError.cullRequestsWhileMoving.cullRequestsWhileMovingMultiplier.preloadWhenHidden.preloadFlightDestinations.preferLeaves.progressiveResolutionHeightFraction.foveatedScreenSpaceError.foveatedConeSize.foveatedMinimumScreenSpaceErrorRelaxation.foveatedInterpolationCallback.foveatedTimeDelay.dynamicScreenSpaceError.dynamicScreenSpaceErrorDensity.dynamicScreenSpaceErrorFactor.dynamicScreenSpaceErrorHeightFalloff.skipLevelOfDetail.baseScreenSpaceError.skipScreenSpaceErrorFactor.skipLevels.immediatelyLoadDesiredLevelOfDetail.loadSiblings.clippingPlanes.clippingPolygons.lightColor.colorBlendAmount.colorBlendMode.debugFreezeFrame.debugColorizeTiles.debugWireframe.debugShowBoundingVolume.debugShowContentBoundingVolume.debugShowViewerRequestVolume.debugShowGeometricError.debugShowRenderingStatistics.debugShowMemoryUsage.debugShowUrl.style.backFaceCulling.showOutline.vectorClassificationOnly.vectorKeepDecodedPositions.splitDirection.customShader.imageBasedLighting.showCreditsOnScreen.featureIdLabel.instanceFeatureIdLabel.outlineColor.cacheBytes.maximumCacheOverflowBytes.enableCollision".split("."),
	cesiumReadonlyProps: [
		"asynchronouslyLoadImagery",
		"classificationType",
		"cullWithChildrenBounds",
		"debugHeatmapTilePropertyName",
		"ellipsoid",
		"enableDebugWireframe",
		"heightReference",
		"modelUpAxis",
		"modelForwardAxis",
		"projectTo2D",
		"enableShowOutline",
		"enablePick",
		"environmentMapOptions",
		"scene"
	],
	cesiumEventProps: {
		onAllTilesLoad: "allTilesLoaded",
		onInitialTilesLoad: "initialTilesLoaded",
		onLoadProgress: "loadProgress",
		onTileFailed: "tileFailed",
		onTileLoad: "tileLoad",
		onTileUnload: "tileUnload",
		onTileVisible: "tileVisible"
	},
	otherProps: [
		"onReady",
		"onError",
		"url"
	],
	useCommonEvent: !0
}), tt = Q({
	name: "Cesium3DTilesetGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new y(t);
		return e.entity.tileset = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.tileset = void 0);
	},
	cesiumProps: [
		"show",
		"uri",
		"maximumScreenSpaceError"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), nt = Q({
	name: "Cesium3DTilesTerrainProvider",
	async create(e, t) {
		let { onReady: n, assetId: r, accessToken: i, url: a, children: o, ...s } = t, c;
		if (r !== void 0) c = await _.fromIonAssetId(r, s);
		else if (a !== void 0) c = await _.fromUrl(a, s);
		else throw Error("Cesium3DTilesTerrainProvider requires either 'url' or 'assetId' prop to be provided");
		return n && n(c), c;
	}
}), rt = Q({
	name: "CesiumWidget",
	create(e, t, n) {
		if (!n) return;
		let r = new b(n, t);
		if (r) return typeof t.resolutionScale == "number" && (r.resolutionScale = t.resolutionScale), [r, new Z(r.scene)];
	},
	destroy(e, t, n, r) {
		r && !r.isDestroyed() && r.destroy(), e.isDestroyed() || e.destroy();
	},
	provide(e, t, n, r) {
		return {
			cesiumWidget: e,
			scene: e.scene,
			camera: e.scene.camera,
			imageryLayerCollection: e.scene.globe.imageryLayers,
			primitiveCollection: e.scene.primitives,
			globe: e.scene.globe,
			__$internal: { onUpdate: n?.onUpdate },
			[Y]: r
		};
	},
	containerProps: ({ id: e, className: t, style: n, full: r, containerProps: i }) => ({
		className: t,
		id: e,
		style: {
			...r ? {
				position: "absolute",
				bottom: "0",
				left: "0",
				right: "0",
				top: "0"
			} : {},
			...n
		},
		...i
	}),
	cesiumProps: [
		"resolutionScale",
		"useDefaultRenderLoop",
		"targetFrameRate",
		"useBrowserRecommendedResolution",
		"allowDataSourcesToSuspendAnimation",
		"trackedEntity",
		"clockTrackedDataSource",
		"terrainProvider",
		"creditDisplay"
	],
	cesiumReadonlyProps: [
		"clock",
		"shouldAnimate",
		"ellipsoid",
		"baseLayer",
		"skyBox",
		"skyAtmosphere",
		"sceneMode",
		"scene3DOnly",
		"orderIndependentTranslucency",
		"mapMode2D",
		"mapProjection",
		"globe",
		"showRenderLoopErrors",
		"automaticallyTrackDataSourceClocks",
		"contextOptions",
		"creditContainer",
		"creditViewport",
		"dataSources",
		"shadows",
		"terrainShadows",
		"terrain",
		"requestRenderMode",
		"maximumRenderTimeChange",
		"msaaSamples",
		"blurActiveElementOnCanvasFocus"
	],
	otherProps: [
		"className",
		"id",
		"style",
		"full",
		"containerProps"
	],
	renderContainer: !0,
	useCommonEvent: !0,
	useRootEvent: !0
}), it = Q({
	name: "ClassificationPrimitive",
	async create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new x(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"classificationType",
		"debugShowBoundingVolume",
		"debugShowShadowVolume",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"compressVertices",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances",
		"vertexCacheOptimize",
		"appearance"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), at = Q({
	name: "Clock",
	create: (e) => e.cesiumWidget?.clock,
	cesiumProps: [
		"canAnimate",
		"clockRange",
		"clockStep",
		"currentTime",
		"multiplier",
		"shouldAnimate",
		"startTime",
		"stopTime"
	],
	cesiumEventProps: {
		onStop: "onStop",
		onTick: "onTick"
	},
	setCesiumPropsAfterCreate: !0
}), ot = Q({
	name: "CloudCollection",
	create: (e) => {
		if (!e.primitiveCollection) return;
		let t = new S();
		return e.primitiveCollection.add(t), t;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide: (e) => ({ cloudCollection: e }),
	cesiumProps: [
		"noiseDetail",
		"noiseOffset",
		"show",
		"debugBillboards",
		"debugEllipsoids"
	],
	setCesiumPropsAfterCreate: !0
}), st = Q({
	name: "CorridorGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new C(t);
		return t.classificationType && (n.classificationType = t.classificationType), e.entity.corridor = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.corridor = void 0);
	},
	cesiumProps: [
		"positions",
		"width",
		"cornerType",
		"height",
		"heightReference",
		"extrudedHeight",
		"extrudedHeightReference",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex",
		"classificationType"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), ct = Q({
	name: "CubeMapPanorama",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new w({
			sources: t.sources,
			show: t.show,
			transform: t.transform,
			credit: t.credit
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	setCesiumPropsAfterCreate: !0,
	cesiumProps: ["show", "sources"],
	cesiumReadonlyProps: ["transform", "credit"]
}), lt = Q({
	name: "CumulusCloud",
	create: (e, t) => e.cloudCollection?.add(t),
	destroy(e, t) {
		t.cloudCollection && !t.cloudCollection.isDestroyed() && t.cloudCollection.remove(e);
	},
	cesiumProps: [
		"show",
		"position",
		"scale",
		"maximumSize",
		"slice",
		"brightness",
		"color"
	]
}), ut = Q({
	name: "CustomDataSource",
	create(e, t) {
		if (!e.dataSourceCollection) return;
		let n = new T(t.name);
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), t.clock !== void 0 && (n.clock = t.clock), e.dataSourceCollection.add(n), n;
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return {
			entityCollection: e.entities,
			dataSource: e
		};
	},
	cesiumProps: [
		"clustering",
		"name",
		"show",
		"clock",
		"isLoading"
	],
	cesiumEventProps: {
		onChange: "changedEvent",
		onError: "errorEvent",
		onLoading: "loadingEvent"
	},
	useCommonEvent: !0
}), dt = Q({
	name: "CylinderGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new E(t);
		return e.entity.cylinder = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.cylinder = void 0);
	},
	cesiumProps: [
		"heightReference",
		"length",
		"topRadius",
		"bottomRadius",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"numberOfVerticalLines",
		"slices",
		"distanceDisplayCondition",
		"shadows"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), ft = ["clustering", "show"], pt = [
	"name",
	"sourceUri",
	"credit"
], mt = {
	onChange: "changedEvent",
	onError: "errorEvent",
	onLoading: "loadingEvent"
}, ht = ["onLoad", "data"], gt = (e, { data: t, onLoad: n, ...r }) => {
	t && e.load(t, r).then((e) => {
		n && n(e);
	});
}, _t = Q({
	name: "CzmlDataSource",
	create(e, t) {
		if (!e.dataSourceCollection) return;
		let n = new D(t.name);
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), e.dataSourceCollection.add(n), t.data && gt(n, t), n;
	},
	update(e, t, n) {
		t.data ? n.show !== t.show && (e.show = typeof t.show == "boolean" ? t.show : !0) : e.show = !1, t.data && (n.data !== t.data || n.sourceUri !== t.sourceUri || n.credit !== t.credit) && gt(e, t);
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return { dataSource: e };
	},
	cesiumProps: ft,
	cesiumReadonlyProps: pt,
	cesiumEventProps: mt,
	otherProps: ht,
	useCommonEvent: !0
}), vt = Q({
	name: "EllipseGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new O(t);
		return e.entity.ellipse = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.ellipse = void 0);
	},
	cesiumProps: [
		"semiMajorAxis",
		"semiMinorAxis",
		"height",
		"heightReference",
		"extrudedHeight",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"numberOfVerticalLines",
		"rotation",
		"stRotation",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex",
		"classificationType",
		"extrudedHeightReference"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), yt = Q({
	name: "EllipsoidGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new k(t);
		return e.entity.ellipsoid = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.ellipsoid = void 0);
	},
	cesiumProps: [
		"heightReference",
		"radii",
		"show",
		"fill",
		"innerRadii",
		"material",
		"maximumClock",
		"maximumCone",
		"minimumClock",
		"minimumCone",
		"outline",
		"outlineColor",
		"outlineWidth",
		"subdivisions",
		"stackPartitions",
		"slicePartitions",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), bt = Q({
	name: "Entity",
	create(e, t) {
		if (!e.entityCollection) return;
		let n = new A(t);
		return e.viewer && t.selected && (e.viewer.selectedEntity = n), e.viewer && t.tracked && (e.viewer.trackedEntity = n), e.entityCollection.add(n), n;
	},
	destroy(e, t) {
		t.entityCollection && t.entityCollection.remove(e);
	},
	update(e, t, n, r) {
		r.viewer && (t.selected !== n.selected && (t.selected ? r.viewer.selectedEntity = e : r.viewer.selectedEntity === e && (r.viewer.selectedEntity = void 0)), t.tracked !== n.tracked && (t.tracked ? r.viewer.trackedEntity = e : r.viewer.trackedEntity === e && (r.viewer.trackedEntity = void 0)));
	},
	provide(e, t, n) {
		return {
			entity: e,
			__$internal: { onUpdate: n?.onUpdate }
		};
	},
	cesiumProps: /* @__PURE__ */ "availability.billboard.box.corridor.cylinder.description.ellipse.ellipsoid.entityCollection.label.model.name.orientation.path.plane.parent.point.polygon.polyline.polylineVolume.position.properties.rectangle.show.tileset.trackingReferenceFrame.viewFrom.wall".split("."),
	cesiumReadonlyProps: ["id"],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" },
	otherProps: ["selected", "tracked"],
	useCommonEvent: !0
}), xt = Q({
	name: "EquirectangularPanorama",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new j({
			transform: t.transform,
			image: t.image,
			radius: t.radius,
			repeatHorizontal: t.repeatHorizontal,
			repeatVertical: t.repeatVertical,
			credit: t.credit
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	setCesiumPropsAfterCreate: !0,
	cesiumProps: ["show"],
	cesiumReadonlyProps: [
		"transform",
		"image",
		"radius",
		"repeatHorizontal",
		"repeatVertical",
		"credit"
	]
}), St = ({ children: e, container: t, resizeInfoBox: n = !0 }) => {
	let { viewer: r, entity: a } = W(), [o, c] = l(!1), u = s(() => t ?? r?.infoBox.frame.contentDocument?.createElement("div"), [t, r?.infoBox.frame.contentDocument]);
	return i(() => {
		if (!r || !a) return;
		let e = (e) => {
			c(!!e && e.id === a.id);
		};
		return r.selectedEntityChanged.addEventListener(e), () => {
			r.selectedEntityChanged.removeEventListener(e);
		};
	}, [a, r]), i(() => {
		if (t || !u || !r) return;
		let e = r.infoBox?.frame, i = e?.contentDocument?.querySelector(".cesium-infoBox-description");
		if (!e || !i) return;
		let a;
		if (o) {
			if (n) {
				let t = i.getBoundingClientRect().height;
				e.style.height = t + "px", a = window.setInterval(() => {
					let t = r.infoBox.container.querySelector(".cesium-infoBox.cesium-infoBox-visible");
					t && (clearInterval(a), a = void 0, i.appendChild(u), t.classList.remove("cesium-infoBox-bodyless"), e.style.height = i.getBoundingClientRect().height + "px");
				}, 10);
			}
		} else u.parentElement === i && i.removeChild(u);
		return a ? () => clearTimeout(a) : void 0;
	}, [
		u,
		t,
		n,
		o,
		r
	]), u ? /* @__PURE__ */ V(B, { children: ke(/* @__PURE__ */ V(B, { children: !t || o ? e : null }), u) }) : null;
}, Ct = Q({
	name: "Fog",
	create(e) {
		if (!e.scene) return;
		let t = new M();
		return e.scene.fog = t, t;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && (t.scene.fog = new M());
	},
	cesiumProps: [
		"density",
		"enabled",
		"heightFalloff",
		"heightScalar",
		"maxHeight",
		"minimumBrightness",
		"renderable",
		"screenSpaceErrorFactor",
		"visualDensityScalar"
	],
	setCesiumPropsAfterCreate: !0
}), wt = [
	"clustering",
	"name",
	"show"
], Tt = [
	"clampToGround",
	"sourceUri",
	"credit",
	"markerSize",
	"markerSymbol",
	"markerColor",
	"stroke",
	"strokeWidth",
	"fill",
	"describe"
], Et = {
	onChange: "changedEvent",
	onError: "errorEvent",
	onLoading: "loadingEvent"
}, Dt = ["onLoad", "data"], Ot = (e, { data: t, onLoad: n, ...r }) => {
	t && e.load(t, r).then((e) => {
		n && n(e);
	});
}, kt = Q({
	name: "GeoJsonDataSource",
	create(e, t) {
		if (!e.dataSourceCollection) return;
		let n = new N(t.name);
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), e.dataSourceCollection.add(n), t.data && Ot(n, t), n;
	},
	update(e, t, n) {
		t.data ? n.show !== t.show && (e.show = typeof t.show == "boolean" ? t.show : !0) : e.show = !1, t.data && (n.data !== t.data || n.clampToGround !== t.clampToGround || n.sourceUri !== t.sourceUri || n.credit !== t.credit || n.markerSize !== t.markerSize || n.markerSymbol !== t.markerSymbol || n.markerColor !== t.markerColor || n.stroke !== t.stroke || n.strokeWidth !== t.strokeWidth || n.fill !== t.fill) && Ot(e, t);
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return { dataSource: e };
	},
	cesiumProps: wt,
	cesiumReadonlyProps: Tt,
	cesiumEventProps: Et,
	otherProps: Dt,
	useCommonEvent: !0
}), At = Q({
	name: "Globe",
	create: (e) => e.scene?.globe,
	update: async (e, t) => {
		let n = t.terrainProvider, r;
		r = J(n) ? await n : n, e.terrainProvider = r;
	},
	cesiumProps: /* @__PURE__ */ "atmosphereBrightnessShift.atmosphereHueShift.atmosphereSaturationShift.backFaceCulling.baseColor.clippingPlanes.depthTestAgainstTerrain.enableLighting.lightingFadeInDistance.lightingFadeOutDistance.material.maximumScreenSpaceError.nightFadeInDistance.nightFadeOutDistance.oceanNormalMapUrl.shadows.show.showGroundAtmosphere.showWaterEffect.tileCacheSize.loadingDescendantLimit.preloadAncestors.preloadSiblings.fillHighlightColor.dynamicAtmosphereLighting.dynamicAtmosphereLightingFromSun.showSkirts.cartographicLimitRectangle.translucency.undergroundColor.undergroundColorAlphaByDistance.lambertDiffuseMultiplier.atmosphereLightIntensity.atmosphereRayleighCoefficient.atmosphereMieCoefficient.atmosphereRayleighScaleHeight.atmosphereMieScaleHeight.atmosphereMieAnisotropy.vertexShadowDarkness.clippingPolygons".split("."),
	cesiumEventProps: {
		onImageryLayersUpdate: "imageryLayersUpdatedEvent",
		onTerrainProviderChange: "terrainProviderChanged",
		onTileLoadProgress: "tileLoadProgressEvent"
	},
	otherProps: ["terrainProvider"],
	setCesiumPropsAfterCreate: !0
}), jt = Q({
	name: "Google2DImageryProvider",
	async create(e, t) {
		let { onReady: n, assetId: r, accessToken: i, key: a, children: o, ...s } = t, c;
		if (r !== void 0) {
			let e = {
				...s,
				assetId: r
			};
			i !== void 0 && (e.accessToken = i), c = await P.fromIonAssetId(e);
		} else {
			let e = { ...s };
			a !== void 0 && (e.key = a), c = await P.fromUrl(e);
		}
		return n && n(c), c;
	}
}), Mt = Q({
	name: "GroundPolylinePrimitive",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new re(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"appearance",
		"classificationType",
		"debugShowBoundingVolume",
		"debugShowShadowVolume",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), Nt = Q({
	name: "GroundPrimitive",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ie(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"appearance",
		"classificationType",
		"debugShowBoundingVolume",
		"debugShowShadowVolume",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"compressVertices",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances",
		"vertexCacheOptimize"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), Pt = Q({
	name: "GroundPrimitiveCollection",
	create: (e) => e.scene?.groundPrimitives,
	provide: (e) => ({ primitiveCollection: e }),
	cesiumProps: [
		"show",
		"destroyPrimitives",
		"primitiveAdded",
		"primitiveRemoved"
	],
	setCesiumPropsAfterCreate: !0
}), Ft = Q({
	name: "ImageryLayer",
	async create(e, t) {
		if (!e.imageryLayerCollection) return;
		let n = J(t.imageryProvider) ? t.imageryProvider : new Promise((e) => queueMicrotask(() => e(t.imageryProvider))), r = e.__$internal?.imageryLayerWaitingList?.slice();
		e.__$internal?.imageryLayerWaitingList && e.__$internal.imageryLayerWaitingList.push(n), r && await Promise.all(r.filter((e) => J(e)));
		let i = await n;
		if (e.__$internal?.imageryLayerWaitingList && (e.__$internal.imageryLayerWaitingList = e.__$internal.imageryLayerWaitingList.filter((e) => e !== n)), !i) return;
		let a = new ae(i, t);
		return e.imageryLayerCollection.add(a, t.index), a;
	},
	destroy(e, t) {
		t.imageryLayerCollection && t.imageryLayerCollection.remove(e);
	},
	cesiumProps: [
		"alpha",
		"brightness",
		"contrast",
		"hue",
		"saturation",
		"gamma",
		"splitDirection",
		"minificationFilter",
		"magnificationFilter",
		"cutoutRectangle",
		"show",
		"nightAlpha",
		"dayAlpha",
		"colorToAlpha",
		"colorToAlphaThreshold",
		"index"
	],
	cesiumReadonlyProps: [
		"rectangle",
		"maximumAnisotropy",
		"minimumTerrainLevel",
		"maximumTerrainLevel",
		"imageryProvider"
	]
}), It = Q({
	name: "ImageryLayerCollection",
	create: (e) => e.globe?.imageryLayers,
	cesiumEventProps: {
		onLayerAdd: "layerAdded",
		onLayerMove: "layerMoved",
		onLayerRemove: "layerRemoved",
		onLayerShowOrHide: "layerShownOrHidden"
	}
}), Lt = [
	"clustering",
	"name",
	"show"
], Rt = [
	"canvas",
	"camera",
	"ellipsoid",
	"clampToGround",
	"sourceUri",
	"credit",
	"screenOverlayContainer"
], zt = {
	onChange: "changedEvent",
	onError: "errorEvent",
	onLoading: "loadingEvent",
	onRefresh: "refreshEvent",
	onUnsupportedNode: "unsupportedNodeEvent"
}, Bt = ["onLoad", "data"], Vt = (e, { data: t, onLoad: n, ...r }) => {
	t && e.load(t, r).then((e) => {
		n && n(e);
	});
}, Ht = Q({
	name: "KmlDataSource",
	create(e, t) {
		if (!e.scene || !e.dataSourceCollection || !e.scene) return;
		let n = new oe({
			camera: t.camera || e.scene.camera,
			canvas: t.canvas || e.scene.canvas,
			ellipsoid: t.ellipsoid,
			credit: t.credit
		});
		return t.clustering && (n.clustering = t.clustering), typeof t.show == "boolean" && (n.show = t.show), t.name !== void 0 && (n.name = t.name), e.dataSourceCollection.add(n), t.data && Vt(n, t), n;
	},
	update(e, t, n) {
		t.data ? n.show !== t.show && (e.show = typeof t.show == "boolean" ? t.show : !0) : e.show = !1, t.data && (n.data !== t.data || n.clampToGround !== t.clampToGround || n.ellipsoid !== t.ellipsoid || n.sourceUri !== t.sourceUri || n.credit !== t.credit) && Vt(e, t);
	},
	destroy(e, t) {
		t.dataSourceCollection && !t.dataSourceCollection.isDestroyed() && t.dataSourceCollection.remove(e);
	},
	provide(e) {
		return { dataSource: e };
	},
	cesiumProps: Lt,
	cesiumReadonlyProps: Rt,
	cesiumEventProps: zt,
	otherProps: Bt,
	useCommonEvent: !0
}), Ut = Q({
	name: "Label",
	create: (e, t) => e.labelCollection?.add(t),
	destroy(e, t) {
		t.labelCollection && !t.labelCollection.isDestroyed() && t.labelCollection.remove(e);
	},
	cesiumProps: [
		"backgroundColor",
		"backgroundPadding",
		"disableDepthTestDistance",
		"distanceDisplayCondition",
		"eyeOffset",
		"fillColor",
		"font",
		"heightReference",
		"horizontalOrigin",
		"id",
		"outlineColor",
		"outlineWidth",
		"pixelOffset",
		"pixelOffsetScaleByDistance",
		"position",
		"scale",
		"scaleByDistance",
		"show",
		"showBackground",
		"style",
		"text",
		"translucencyByDistance",
		"verticalOrigin"
	],
	useCommonEvent: !0
}), Wt = Q({
	name: "LabelCollection",
	create(e, t) {
		if (!e.scene || !e.primitiveCollection) return;
		let n = new se({
			scene: e.scene,
			modelMatrix: t.modelMatrix,
			blendOption: t.blendOption,
			debugShowBoundingVolume: t.debugShowBoundingVolume
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { labelCollection: e };
	},
	cesiumProps: [
		"blendOption",
		"coarseDepthTestDistance",
		"debugShowBoundingVolume",
		"modelMatrix",
		"show",
		"threePointDepthTestDistance"
	],
	setCesiumPropsAfterCreate: !0
}), Gt = Q({
	name: "LabelGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new ce(t);
		return e.entity.label = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.label = void 0);
	},
	cesiumProps: [
		"text",
		"font",
		"style",
		"fillColor",
		"outlineColor",
		"outlineWidth",
		"show",
		"showBackground",
		"backgroundColor",
		"backgroundPadding",
		"scale",
		"horizontalOrigin",
		"verticalOrigin",
		"eyeOffset",
		"pixelOffset",
		"translucencyByDistance",
		"pixelOffsetScaleByDistance",
		"scaleByDistance",
		"heightReference",
		"distanceDisplayCondition",
		"disableDepthTestDistance"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Kt = Q({
	name: "Model",
	async create(e, { scene: t, url: n, colorBlendMode: r, ...i }) {
		if (!e.scene || !e.primitiveCollection || !n) return;
		let a = n, o;
		o = J(a) ? await a : a;
		let s;
		try {
			s = await le.fromGltfAsync({
				...i,
				url: o,
				colorBlendMode: r,
				scene: t || e.scene
			});
		} catch (e) {
			i.onError?.(e);
			return;
		}
		return e.primitiveCollection.add(s), s;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumEventProps: { onReady: "readyEvent" },
	cesiumProps: /* @__PURE__ */ "backFaceCulling.clampAnimations.clippingPlanes.clippingPolygons.color.colorBlendAmount.colorBlendMode.customShader.debugShowBoundingVolume.debugWireframe.distanceDisplayCondition.enableVerticalExaggeration.featureIdLabel.heightReference.id.imageBasedLighting.instanceFeatureIdLabel.lightColor.maximumScale.minimumPixelSize.modelMatrix.outlineColor.scale.shadows.show.showCreditsOnScreen.showOutline.silhouetteColor.silhouetteSize.splitDirection.style.pointCloudShading".split("."),
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"basePath",
		"credit",
		"enableDebugWireframe",
		"gltf",
		"incrementallyLoadTextures",
		"scene",
		"releaseGltfJson",
		"cull",
		"opaquePass",
		"upAxis",
		"forwardAxis",
		"content",
		"enableShowOutline",
		"projectTo2D",
		"classificationType",
		"gltfCallback",
		"enablePick",
		"environmentMapOptions"
	],
	otherProps: [
		"onReady",
		"onError",
		"url"
	],
	useCommonEvent: !0
}), qt = Q({
	name: "ModelGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new ue(t);
		return e.entity.model = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.model = void 0);
	},
	cesiumProps: [
		"uri",
		"show",
		"scale",
		"minimumPixelSize",
		"maximumScale",
		"incrementallyLoadTextures",
		"runAnimations",
		"clampAnimations",
		"nodeTransformations",
		"shadows",
		"heightReference",
		"distanceDisplayCondition",
		"silhouetteColor",
		"silhouetteSize",
		"color",
		"colorBlendMode",
		"colorBlendAmount",
		"clippingPlanes",
		"imageBasedLightingFactor",
		"lightColor",
		"articulations",
		"customShader",
		"enableVerticalExaggeration",
		"environmentMapOptions"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Jt = Q({
	name: "Moon",
	create(e, t) {
		if (!e.scene) return;
		let n = new de(t);
		return e.scene.moon = n, n;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && (t.scene.moon = new de());
	},
	cesiumProps: [
		"onlySunLighting",
		"show",
		"textureUrl"
	],
	cesiumReadonlyProps: ["ellipsoid"]
}), Yt = Q({
	name: "ParticleSystem",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new fe({
			...t,
			updateCallback: t.onUpdate
		});
		return e.primitiveCollection.add(n), n;
	},
	update(e, t, n) {
		t.onUpdate !== n.onUpdate && (e.updateCallback = t.onUpdate);
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e);
	},
	cesiumProps: /* @__PURE__ */ "show.emitter.modelMatrix.emitterModelMatrix.emissionRate.bursts.loop.startScale.endScale.startColor.endColor.image.imageSize.minimumImageSize.maximumImageSize.speed.minimumSpeed.maximumSpeed.lifetime.particleLife.minimumParticleLife.maximumParticleLife.mass.minimumMass.maximumMass.sizeInMeters".split("."),
	cesiumEventProps: {
		onComplete: "complete",
		onUpdate: "updateCallback"
	}
}), Xt = Q({
	name: "PathGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new pe(t);
		return e.entity.path = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.path = void 0);
	},
	cesiumProps: [
		"leadTime",
		"trailTime",
		"show",
		"width",
		"material",
		"resolution",
		"distanceDisplayCondition",
		"relativeTo"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Zt = Q({
	name: "PlaneGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new me(t);
		return e.entity.plane = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.plane = void 0);
	},
	cesiumProps: [
		"plane",
		"dimensions",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Qt = Q({
	name: "PointGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new he(t);
		return e.entity.point = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.point = void 0);
	},
	cesiumProps: [
		"color",
		"pixelSize",
		"outlineColor",
		"outlineWidth",
		"show",
		"scaleByDistance",
		"translucencyByDistance",
		"heightReference",
		"distanceDisplayCondition",
		"disableDepthTestDistance",
		"splitDirection"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), $t = Q({
	name: "PointPrimitive",
	create: (e, t) => e.pointPrimitiveCollection?.add(t),
	destroy(e, t) {
		t.pointPrimitiveCollection && !t.pointPrimitiveCollection.isDestroyed() && t.pointPrimitiveCollection.remove(e);
	},
	cesiumProps: [
		"color",
		"disableDepthTestDistance",
		"distanceDisplayCondition",
		"id",
		"outlineColor",
		"outlineWidth",
		"pixelSize",
		"position",
		"scaleByDistance",
		"show",
		"splitDirection",
		"translucencyByDistance"
	],
	useCommonEvent: !0
}), en = Q({
	name: "PointPrimitveCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ge(t);
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { pointPrimitiveCollection: e };
	},
	cesiumProps: [
		"blendOption",
		"debugShowBoundingVolume",
		"modelMatrix",
		"show"
	]
}), tn = Q({
	name: "PolygonGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new _e(t);
		return e.entity.polygon = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.polygon = void 0);
	},
	cesiumProps: [
		"arcType",
		"hierarchy",
		"height",
		"heightReference",
		"extrudedHeight",
		"extrudedHeightReference",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"stRotation",
		"granularity",
		"perPositionHeight",
		"closeTop",
		"closeBottom",
		"shadows",
		"distanceDisplayCondition",
		"zIndex",
		"classificationType",
		"textureCoordinates"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), nn = Q({
	name: "Polyline",
	create: (e, t) => e.polylineCollection?.add(t),
	destroy(e, t) {
		t.polylineCollection && !t.polylineCollection.isDestroyed() && t.polylineCollection.remove(e);
	},
	cesiumProps: [
		"distanceDisplayCondition",
		"id",
		"loop",
		"material",
		"positions",
		"show",
		"width"
	],
	useCommonEvent: !0
}), rn = Q({
	name: "PolylineCollection",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new ve({
			modelMatrix: t.modelMatrix,
			debugShowBoundingVolume: t.debugShowBoundingVolume,
			length: t.length,
			scene: e.scene
		});
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { polylineCollection: e };
	},
	cesiumProps: [
		"debugShowBoundingVolume",
		"length",
		"modelMatrix",
		"show"
	]
}), an = Q({
	name: "PolylineGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new ye(t);
		return e.entity.polyline = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.polyline = void 0);
	},
	cesiumProps: [
		"arcType",
		"classificationType",
		"positions",
		"clampToGround",
		"width",
		"show",
		"material",
		"depthFailMaterial",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), on = Q({
	name: "PolylineVolumeGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new be(t);
		return e.entity.polylineVolume = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.polylineVolume = void 0);
	},
	cesiumProps: [
		"positions",
		"shape",
		"cornerType",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"granularity",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), sn = Q({
	name: "PostProcessStage",
	create(e, t) {
		if (!e.scene) return;
		let n = new xe(t);
		return typeof t.enabled == "boolean" && (n.enabled = t.enabled), t.selected && (n.selected = t.selected), e.scene.postProcessStages.add(n), n;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && t.scene.postProcessStages.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: ["enabled", "selected"],
	cesiumReadonlyProps: [
		"clearColor",
		"forcePowerOfTwo",
		"fragmentShader",
		"name",
		"pixelDatatype",
		"pixelFormat",
		"sampleMode",
		"scissorRectangle",
		"textureScale",
		"uniforms"
	]
}), cn = $({
	name: "BlackAndWhiteStage",
	props: ["gradations"],
	create: () => F.createBlackAndWhiteStage()
}), ln = $({
	name: "BrightnessStage",
	props: ["brightness"],
	create: () => F.createBrightnessStage()
}), un = $({
	name: "LensFlareStage",
	props: [
		"dirtTexture",
		"starTexture",
		"intensity",
		"distortion",
		"ghostDispersal",
		"haloWidth",
		"earthRadius"
	],
	create: () => F.createLensFlareStage()
}), dn = $({
	name: "Fxaa",
	create: (e, t) => t.fxaa,
	props: []
}), fn = $({
	name: "NightVisionStage",
	props: [],
	create: () => F.createNightVisionStage()
}), pn = Q({
	name: "PostProcessStageComposite",
	create(e, t) {
		if (!e.scene) return;
		let n = new Se(t);
		return typeof t.enabled == "boolean" && (n.enabled = t.enabled), t.selected && (n.selected = t.selected), e.scene.postProcessStages.add(n), n;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && t.scene.postProcessStages.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: ["enabled", "selected"],
	cesiumReadonlyProps: [
		"inputPreviousStageTexture",
		"name",
		"stages",
		"uniforms"
	]
}), mn = $({
	name: "AmbientOcclusion",
	create: (e, t) => t.ambientOcclusion,
	props: [
		"ambientOcclusionOnly",
		"bias",
		"delta",
		"frustumLength",
		"intensity",
		"lengthCap",
		"sigma",
		"stepSize"
	],
	noMount: !0
}), hn = $({
	name: "Bloom",
	create: (e, t) => t.bloom,
	props: [
		"brightness",
		"contrast",
		"delta",
		"glowOnly",
		"sigma",
		"stepSize"
	],
	noMount: !0
}), gn = $({
	name: "BlurStage",
	props: [
		"delta",
		"sigma",
		"stepSize"
	],
	create: () => F.createBlurStage()
}), _n = $({
	name: "DepthOfFieldStage",
	props: [
		"delta",
		"focalDistance",
		"sigma",
		"stepSize"
	],
	create: () => F.createDepthOfFieldStage()
}), vn = $({
	name: "EdgeDetectionStage",
	props: ["color", "length"],
	create: () => F.createEdgeDetectionStage()
}), yn = $({
	name: "SilhouetteStage",
	props: ["color", "length"],
	create: () => F.createSilhouetteStage()
}), bn = Q({
	name: "Primitive",
	create(e, t) {
		if (!e.primitiveCollection) return;
		let n = new Ce(t);
		if (t.onReady) {
			let r = () => {
				n.ready && (t.onReady?.(n), e.scene?.postRender.removeEventListener(r));
			};
			e.scene?.postRender.addEventListener(r);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"appearance",
		"cull",
		"debugShowBoundingVolume",
		"depthFailAppearance",
		"modelMatrix",
		"shadows",
		"show"
	],
	cesiumReadonlyProps: [
		"allowPicking",
		"asynchronous",
		"compressVertices",
		"geometryInstances",
		"interleave",
		"releaseGeometryInstances",
		"vertexCacheOptimize"
	],
	otherProps: ["onReady"],
	useCommonEvent: !0
}), xn = Q({
	name: "RectangleGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new we(t);
		return e.entity.rectangle = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.rectangle = void 0);
	},
	cesiumProps: [
		"classificationType",
		"coordinates",
		"height",
		"heightReference",
		"extrudedHeight",
		"extrudedHeightReference",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"rotation",
		"stRotation",
		"granularity",
		"shadows",
		"distanceDisplayCondition",
		"zIndex"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
}), Sn = /* @__PURE__ */ "backgroundColor.completeMorphOnUserInput.debugShowCommands.debugShowDepthFrustum.debugShowFramesPerSecond.debugShowFrustumPlanes.debugShowFrustums.eyeSeparation.farToNearRatio.focalLength.fog.gamma.globe.highDynamicRange.invertClassification.invertClassificationColor.light.logarithmicDepthBuffer.logarithmicDepthFarToNearRatio.maximumRenderTimeChange.minimumDisableDepthTestDistance.moon.morphTime.nearToFarDistance2D.pickTranslucentDepth.requestRenderMode.rethrowRenderErrors.shadowMap.skyAtmosphere.skyBox.specularEnvironmentMaps.sphericalHarmonicCoefficients.sun.sunBloom.terrainProvider.useDepthPicking.useWebVR.msaaSamples.splitPosition.verticalExaggeration.verticalExaggerationRelativeHeight.atmosphere".split("."), Cn = {
	onMorphComplete: "morphComplete",
	onMorphStart: "morphStart",
	onPostRender: "postRender",
	onPreRender: "preRender",
	onPreUpdate: "preUpdate",
	onPostUpdate: "postUpdate",
	onRenderError: "renderError",
	onTerrainProviderChange: "terrainProviderChanged"
}, wn = [
	"mode",
	"morphDuration",
	"enableEdgeVisibility"
], Tn = (e, t, n) => {
	switch (t) {
		case I.SCENE2D:
			e.morphTo2D(n);
			break;
		case I.COLUMBUS_VIEW:
			e.morphToColumbusView(n);
			break;
		case I.SCENE3D:
			e.morphTo3D(n);
			break;
	}
}, En = Q({
	name: "Scene",
	create(e, t) {
		return e.scene && (t.mode && Tn(e.scene, t.mode, t.morphDuration), t.enableEdgeVisibility !== void 0 && (e.scene._enableEdgeVisibility = t.enableEdgeVisibility)), e.scene;
	},
	update(e, t, n) {
		t.mode !== n.mode && t.mode && Tn(e, t.mode, t.morphDuration), t.enableEdgeVisibility !== n.enableEdgeVisibility && t.enableEdgeVisibility !== void 0 && (e._enableEdgeVisibility = t.enableEdgeVisibility);
	},
	cesiumProps: Sn,
	cesiumEventProps: Cn,
	otherProps: wn,
	setCesiumPropsAfterCreate: !0
}), Dn = Q({
	name: "ScreenSpaceCameraController",
	create: (e) => e.scene?.screenSpaceCameraController,
	cesiumProps: [
		"bounceAnimationTime",
		"enableCollisionDetection",
		"enableInputs",
		"enableLook",
		"enableRotate",
		"enableTilt",
		"enableTranslate",
		"enableZoom",
		"inertiaSpin",
		"inertiaTranslate",
		"inertiaZoom",
		"lookEventTypes",
		"maximumMovementRatio",
		"maximumZoomDistance",
		"minimumCollisionTerrainHeight",
		"minimumPickingTerrainHeight",
		"minimumTrackBallHeight",
		"minimumZoomDistance",
		"rotateEventTypes",
		"tiltEventTypes",
		"translateEventTypes",
		"zoomEventTypes",
		"minimumPickingTerrainDistanceWithInertia",
		"maximumTiltAngle",
		"zoomFactor"
	],
	setCesiumPropsAfterCreate: !0
}), On = ({ action: e, modifier: t, type: n }) => {
	let r = W();
	return i(() => {
		if (!(!r.screenSpaceEventHandler || r.screenSpaceEventHandler.isDestroyed())) {
			if (e) return r.screenSpaceEventHandler.setInputAction(e, n, t), () => {
				!r.screenSpaceEventHandler || r.screenSpaceEventHandler.isDestroyed() || r.screenSpaceEventHandler.removeInputAction(n, t);
			};
			r.screenSpaceEventHandler.removeInputAction(n, t);
		}
	}, [
		e,
		r.screenSpaceEventHandler,
		t,
		n
	]), null;
}, kn = Q({
	name: "ScreenSpaceEventHandler",
	create(e, t) {
		return t.useDefault ? e.cesiumWidget?.screenSpaceEventHandler : e.scene ? new L(e.scene.canvas) : void 0;
	},
	destroy(e) {
		e.isDestroyed() || e.destroy();
	},
	provide(e) {
		return { screenSpaceEventHandler: e };
	}
}), An = Q({
	name: "ShadowMap",
	create: (e) => e.scene?.shadowMap,
	cesiumProps: [
		"darkness",
		"fadingEnabled",
		"maximumDistance",
		"enabled",
		"normalOffset",
		"softShadows",
		"size"
	]
}), jn = Q({
	name: "SkyAtmosphere",
	create: (e) => e.scene?.skyAtmosphere,
	cesiumProps: [
		"brightnessShift",
		"hueShift",
		"saturationShift",
		"show",
		"perFragmentAtmosphere",
		"atmosphereLightIntensity",
		"atmosphereRayleighCoefficient",
		"atmosphereMieCoefficient",
		"atmosphereRayleighScaleHeight",
		"atmosphereMieScaleHeight",
		"atmosphereMieAnisotropy"
	],
	setCesiumPropsAfterCreate: !0
}), Mn = Q({
	name: "SkyBox",
	create: (e) => e.scene?.skyBox,
	cesiumProps: ["sources", "show"],
	setCesiumPropsAfterCreate: !0
}), Nn = Q({
	name: "Sun",
	create(e) {
		if (!e.scene) return;
		let t = new z();
		return e.scene.sun = t, t;
	},
	destroy(e, t) {
		t.scene && !t.scene.isDestroyed() && (t.scene.sun = new z());
	},
	cesiumProps: ["glowFactor", "show"],
	setCesiumPropsAfterCreate: !0
}), Pn = Q({
	name: "TimeDynamicPointCloud",
	create(e, t) {
		if (!e.cesiumWidget || !e.primitiveCollection || !e.cesiumWidget?.clock) return;
		let n = new Te({
			...t,
			clock: t.clock ?? e.cesiumWidget.clock
		});
		if (t.onReady) {
			let e = () => {
				t.onReady?.(n), n.frameChanged.removeEventListener(e);
			};
			n.frameChanged.addEventListener(e);
		}
		return e.primitiveCollection.add(n), n;
	},
	destroy(e, t) {
		t.primitiveCollection && !t.primitiveCollection.isDestroyed() && t.primitiveCollection.remove(e), e.isDestroyed() || e.destroy();
	},
	cesiumProps: [
		"clippingPlanes",
		"maximumMemoryUsage",
		"modelMatrix",
		"shadows",
		"show",
		"style",
		"intervals"
	],
	cesiumReadonlyProps: ["clock", "shading"],
	cesiumEventProps: { onFrameChange: "frameChanged" },
	otherProps: ["onReady"],
	useCommonEvent: !0
}), Fn = Q({
	name: "Viewer",
	async create(e, { baseLayer: t, terrainProvider: n, ...r }, i) {
		if (!i) return;
		let a;
		a = J(n) ? await n : n;
		let o = new Ee(i, {
			...r,
			terrainProvider: a,
			baseLayer: t ?? void 0
		});
		if (o) return t === !1 && o.imageryLayers.removeAll(), o && r.extend && (Array.isArray(r.extend) ? r.extend.forEach((e) => {
			o.extend(e, {});
		}) : o.extend(r.extend, {})), [o, new Z(o.scene)];
	},
	destroy(e, t, n, r) {
		r && !r.isDestroyed() && r.destroy(), e.isDestroyed() || e.destroy();
	},
	provide(e, t, n, r) {
		return {
			viewer: e,
			cesiumWidget: e.cesiumWidget,
			dataSourceCollection: e.dataSources,
			entityCollection: e.entities,
			scene: e.scene,
			camera: e.scene.camera,
			imageryLayerCollection: e.scene.globe.imageryLayers,
			primitiveCollection: e.scene.primitives,
			globe: e.scene.globe,
			__$internal: {
				onUpdate: n?.onUpdate,
				imageryLayerWaitingList: []
			},
			[Y]: r
		};
	},
	containerProps: ({ id: e, className: t, style: n, full: r, containerProps: i }) => ({
		className: t,
		id: e,
		style: {
			...r ? {
				position: "absolute",
				bottom: "0",
				left: "0",
				right: "0",
				top: "0"
			} : {},
			...n
		},
		...i
	}),
	cesiumProps: [
		"terrainShadows",
		"clockTrackedDataSource",
		"targetFrameRate",
		"useDefaultRenderLoop",
		"resolutionScale",
		"allowDataSourcesToSuspendAnimation",
		"trackedEntity",
		"selectedEntity",
		"shadows",
		"useBrowserRecommendedResolution",
		"creditDisplay"
	],
	cesiumReadonlyProps: /* @__PURE__ */ "baseLayer.animation.baseLayerPicker.fullscreenButton.vrButton.geocoder.homeButton.infoBox.sceneModePicker.selectionIndicator.timeline.navigationHelpButton.navigationInstructionsInitiallyVisible.scene3DOnly.shouldAnimate.clockViewModel.selectedImageryProviderViewModel.imageryProviderViewModels.selectedTerrainProviderViewModel.terrainProviderViewModels.ellipsoid.skyBox.skyAtmosphere.fullscreenElement.showRenderLoopErrors.automaticallyTrackDataSourceClocks.contextOptions.sceneMode.mapProjection.globe.orderIndependentTranslucency.creditContainer.creditViewport.dataSources.mapMode2D.projectionPicker.blurActiveElementOnCanvasFocus.requestRenderMode.maximumRenderTimeChange.depthPlaneEllipsoidOffset.msaaSamples.terrain".split("."),
	cesiumEventProps: {
		onSelectedEntityChange: "selectedEntityChanged",
		onTrackedEntityChange: "trackedEntityChanged"
	},
	otherProps: [
		"className",
		"id",
		"style",
		"full",
		"containerProps",
		"extend",
		"terrainProvider"
	],
	renderContainer: !0,
	useCommonEvent: !0,
	useRootEvent: !0
}), In = Q({
	name: "WebMapTileServiceImageryProvider",
	create(e, t) {
		let { onReady: n, children: r, ...i } = t, a = new Oe(i);
		return n && n(a), a;
	}
}), Ln = Q({
	name: "WallGraphics",
	create(e, t) {
		if (!e.entity) return;
		let n = new De(t);
		return e.entity.wall = n, n;
	},
	destroy(e, t) {
		t.entity && (t.entity.wall = void 0);
	},
	cesiumProps: [
		"positions",
		"maximumHeights",
		"minimumHeights",
		"show",
		"fill",
		"material",
		"outline",
		"outlineColor",
		"outlineWidth",
		"granularity",
		"shadows",
		"distanceDisplayCondition"
	],
	cesiumEventProps: { onDefinitionChange: "definitionChanged" }
});
//#endregion
export { mn as AmbientOcclusion, ze as Billboard, Be as BillboardCollection, Ve as BillboardGraphics, cn as BlackAndWhiteStage, hn as Bloom, gn as BlurStage, He as BoxGraphics, ln as BrightnessStage, Ue as BufferPoint, We as BufferPointCollection, Ge as BufferPolygon, Ke as BufferPolygonCollection, qe as BufferPolyline, Je as BufferPolylineCollection, Ye as Camera, Xe as CameraFlyHome, Ze as CameraFlyTo, $e as CameraFlyToBoundingSphere, Qe as CameraLookAt, nt as Cesium3DTilesTerrainProvider, et as Cesium3DTileset, tt as Cesium3DTilesetGraphics, H as CesiumContext, rt as CesiumWidget, it as ClassificationPrimitive, at as Clock, ot as CloudCollection, U as Consumer, st as CorridorGraphics, ct as CubeMapPanorama, lt as CumulusCloud, ut as CustomDataSource, dt as CylinderGraphics, _t as CzmlDataSource, _n as DepthOfFieldStage, vn as EdgeDetectionStage, vt as EllipseGraphics, yt as EllipsoidGraphics, bt as Entity, St as EntityDescription, xt as EquirectangularPanorama, Z as EventManager, Ct as Fog, dn as Fxaa, kt as GeoJsonDataSource, At as Globe, jt as Google2DImageryProvider, Mt as GroundPolylinePrimitive, Nt as GroundPrimitive, Pt as GroundPrimitiveCollection, Ft as ImageryLayer, It as ImageryLayerCollection, Ht as KmlDataSource, Ut as Label, Wt as LabelCollection, Gt as LabelGraphics, un as LensFlareStage, Kt as Model, qt as ModelGraphics, Jt as Moon, fn as NightVisionStage, Yt as ParticleSystem, Xt as PathGraphics, Zt as PlaneGraphics, Qt as PointGraphics, $t as PointPrimitive, en as PointPrimitiveCollection, tn as PolygonGraphics, nn as Polyline, rn as PolylineCollection, an as PolylineGraphics, on as PolylineVolumeGraphics, sn as PostProcessStage, pn as PostProcessStageComposite, bn as Primitive, Ae as Provider, xn as RectangleGraphics, En as Scene, Dn as ScreenSpaceCameraController, On as ScreenSpaceEvent, kn as ScreenSpaceEventHandler, An as ShadowMap, yn as SilhouetteStage, jn as SkyAtmosphere, Mn as SkyBox, Nn as Sun, Pn as TimeDynamicPointCloud, Fn as Viewer, Ln as WallGraphics, In as WebMapTileServiceImageryProvider, G as createCameraOperation, Q as createCesiumComponent, $ as createPostProcessStage, K as entries, Y as eventManagerContextKey, X as eventNames, q as includes, Ne as isDestroyable, Pe as isDestroyed, J as isPromise, je as pick, Me as shallowEquals, W as useCesium, Fe as useCesiumComponent, Re as withCesium };
